src/test/java/tel/schich/javacan/test/BcmCanSocketTest.java

```diff
@@ -0,0 +1,205 @@
+        byte[][] frameData = new byte[][] {
+                new byte[] { 1, 2, },
+                new byte[] { 3, 4, 5 },
+                new byte[] { 6, 7, 8, 9 },
+        };
+        BcmMessage message = BcmMessage.builder()
+                .opcode(BcmOpcode.TX_SETUP)
+                .frame(CanFrame.create(0, (byte) 0, frameData[0]))
+                .frame(CanFrame.create(0, (byte) 0, frameData[1]))
+                .frame(CanFrame.create(0, (byte) 0, frameData[2]))
+                .build();
+
+        Iterator<CanFrame> frameIterator = message.getFrames().iterator();
+        // check all frames access
+        for (byte[] expectedData : frameData) {
+            CanFrame frame = frameIterator.next();
+            ByteBuffer dataBuffer = ByteBuffer.allocate(frame.getDataLength());
+            frame.getData(dataBuffer);
+            assertArrayEquals(expectedData, dataBuffer.array());
+        }
+        // check single frames access
+        for (int i = 0; i < frameData.length; i++) {
+            CanFrame frame = message.getFrame(i);
+            ByteBuffer dataBuffer = ByteBuffer.allocate(frame.getDataLength());
+            frame.getData(dataBuffer);
+            assertArrayEquals(frameData[i], dataBuffer.array());
+        }
+    }
+
+    @Test
+    void testMessageFDFrameAccess() {
+        byte[][] frameData = new byte[][] {
+                new byte[] { 1, 2, },
+                new byte[] { 3, 4, 5, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 },
+                new byte[] { 6, 7, 8, 9 },
+        };
+        int expectedBufferSize = BcmMessage.HEADER_LENGTH + frameData.length * RawCanChannel.FD_MTU;
+        BcmMessage message = BcmMessage.builder()
+                .opcode(BcmOpcode.TX_SETUP)
+                .frame(CanFrame.create(0, (byte) 0, frameData[0]))
+                .frame(CanFrame.create(0, (byte) 0, frameData[1]))
+                .frame(CanFrame.create(0, (byte) 0, frameData[2]))
+                .build();
+
+        assertEquals(expectedBufferSize, message.getBuffer().remaining());
+        assertTrue(message.getFlags().contains(BcmFlag.CAN_FD_FRAME));
+
+        Iterator<CanFrame> frameIterator = message.getFrames().iterator();
+        // check all frames access
+        for (byte[] expectedData : frameData) {
+            CanFrame frame = frameIterator.next();
+            ByteBuffer dataBuffer = ByteBuffer.allocate(frame.getDataLength());
+            frame.getData(dataBuffer);
+            assertArrayEquals(expectedData, dataBuffer.array());
+        }
+        // check single frames access
+        for (int i = 0; i < frameData.length; i++) {
+            CanFrame frame = message.getFrame(i);
+            ByteBuffer dataBuffer = ByteBuffer.allocate(frame.getDataLength());
+            frame.getData(dataBuffer);
+            assertArrayEquals(frameData[i], dataBuffer.array());
+        }
+    }
+
+    @Test
+    void testNonBlockingRead() throws Exception {
```

Comment:
without actually looking into the logic, this seems to test more then just a non-blocking read. Consider renaming.

Reply:
i did split the test